### PR TITLE
quiche: make quic proof source and crypto stream pluggable

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: make quic proof source and crypto streams extensions. Add config for default ones. If not specified in config, the default ones will be used.

Risk Level: low
Testing: existing tests passed
Part of #2557

### CODE REVIEW

File: api/envoy/config/listener/v3/quic_config.proto [hunk 1]
```diff
@@ -19,8 +20,22 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: QUIC listener config]
 
+// Configuration specific to QUIC crypto stream.
+message QuicCryptoStream {
+  string name = 1 [(validate.rules).string = {min_len: 1}];
+
+  google.protobuf.Any typed_config = 2;
```
Comment (htuch) [inline api/envoy/config/listener/v3/quic_config.proto:27]: @phlax can you comment on the extension point documentation? Thanks.

File: api/envoy/config/listener/v3/quic_config.proto [hunk 2]
```diff
@@ -48,4 +63,12 @@ message QuicProtocolOptions {
   // bound by 6000, regardless of this field or how many connections there are.
   google.protobuf.UInt32Value packets_to_read_to_connection_count_ratio = 5
       [(validate.rules).uint32 = {gte: 1}];
+
+  // The crypto server stream implementation used for this listener.
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto]: Yes, added quic class names instead.

File: api/envoy/config/listener/v3/quic_config.proto [hunk 3]
```diff
@@ -48,4 +63,12 @@ message QuicProtocolOptions {
...
+  // If not specified the :ref:`QUICHE defaul one<envoy_v3_api_field_extensions.quic.v3.CryptoServerStreamConfig>` will be used.
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto]: fixed

File: api/envoy/config/listener/v3/quic_config.proto [hunk 4]
```diff
@@ -19,8 +20,24 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+  // [#extension-category: envoy.quic.server.crypto_stream]
+  google.protobuf.Any typed_config = 2;
+}
+
+// Configuration specific to QUIC proof source.
+message QuicProofSource {
+  string name = 1 [(validate.rules).string = {min_len: 1}];
+
+  // [#extension-category: envoy.quic.proof_source]
+  google.protobuf.Any typed_config = 2;
```
Comment (htuch) [inline api/envoy/config/listener/v3/quic_config.proto:36]: Why doesn't this use `TypedExtensionConfig`? (and above)

File: api/envoy/config/listener/v3/quic_config.proto [hunk 5]
```diff
@@ -48,4 +65,12 @@ message QuicProtocolOptions {
   // bound by 6000, regardless of this field or how many connections there are.
   google.protobuf.UInt32Value packets_to_read_to_connection_count_ratio = 5
       [(validate.rules).uint32 = {gte: 1}];
+
+  // The implementation of quic::QuicCryptoClientStreamBase used for this listener.
+  // If not specified the :ref:`QUICHE default one<envoy_v3_api_msg_extensions.quic.v3.CryptoServerStreamConfig>` will be used.
+  QuicCryptoStream crypto_stream = 6;
```
Comment (RyanTheOptimist) [inline api/envoy/config/listener/v3/quic_config.proto:71]: I'm not super familiar with the way envoy uses protos to configure different behavior, so I'm sorry for being confused. The comment here talks about an implementation of `quic::QuicCryptoClientStreamBase`, which is a C++ class. This makes me think that the crypto_stream field is going to specify a C++ class. However, the default value for this field is `envoy_v3_api_msg_extensions.quic.v3.CryptoServerStreamConfig`, which is a proto message. What is the mapping between messages and C++ classes? I suspect it must have something to do with getAndCheckFactory()?

File: api/envoy/config/listener/v3/quic_config.proto (hunk 4 above)
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto:36]: I didn't know about it. Switched now.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 5 above)
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto:71]: Hmm, the comment is not supper clear. Fixed now.

Envoy extensions are c++ classes like quic::QuicCryptoClientStreamBase. But how we plug them in is via a 1:1 mapped protobuf. I'm trying to configure which c++ classes to use for quic crypto stream and proof source via the configuration of protobuf messages here.

Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto]: And this is handled via static linked-in library.  envoy/registry/registry.h provides interface for that, and getAndCheckFactory() is a wrapper around those interfaces.

Comment (RyanTheOptimist) [reply api/envoy/config/listener/v3/quic_config.proto]: AH! Thanks, that helps a lot.

Well, other than that the classes in that file read like a randomly generated strings of design patterns :)
* BaseFactoryCategoryRegistry 
* FactoryRegistryProxy
*  InjectFactory

**Review by RyanTheOptimist — DISMISSED** [review]

File: source/common/quic/client_connection_factory_impl.h [hunk 6]
```diff
@@ -48,6 +49,8 @@ struct PersistentQuicInfoImpl : public Http::PersistentQuicInfo {
   // This arguably should not be shared across connections but as Envoy doesn't
   // support push promise it's really moot point.
   quic::QuicClientPushPromiseIndex push_promise_index_;
+  // Hard code with real crypto stream for now.
```
Comment (alyssawilk) [inline source/common/quic/client_connection_factory_impl.h:52]: Please either add a TODO, or change the API docs to make it clear it's downstream only and config validate.

File: source/extensions/extensions_metadata.yaml [hunk 7]
```diff
@@ -489,6 +489,16 @@ envoy.matching.input_matchers.consistent_hashing:
...
   status: stable
+envoy.quic.filter_chain_proof_source:
+  categories:
+  - envoy.quic.proof_source
+  security_posture: robust_to_untrusted_downstream
+  status: alpha
+quic.quiche_crypto_server_stream:
+  categories:
+  - envoy.quic.server.crypto_stream
+  security_posture: robust_to_untrusted_downstream
```
Comment (alyssawilk) [inline source/extensions/extensions_metadata.yaml:500]: I'm not sure on posture here, given alpha.  @yanavlasov for comments

File: test/common/quic/envoy_quic_client_session_test.cc [hunk 8]
```diff
@@ -74,22 +77,17 @@ class TestQuicCryptoClientStream : public quic::QuicCryptoClientStream {
...
-    return std::make_unique<TestQuicCryptoClientStream>(
-        server_id(), this, crypto_config()->proof_verifier()->CreateDefaultContext(),
-        crypto_config(), this, true);
+  std::unique_ptr<quic::QuicCryptoClientStreamBase>
+  createEnvoyQuicCryptoClientStream(const quic::QuicServerId& server_id, quic::QuicSession* session,
+                                    std::unique_ptr<quic::ProofVerifyContext> verify_context,
+                                    quic::QuicCryptoClientConfig* crypto_config,
+                                    quic::QuicCryptoClientStream::ProofHandler* proof_handler,
+                                    bool has_application_state) override {
+    return std::make_unique<TestQuicCryptoClientStream>(server_id, session,
```
Comment (alyssawilk) [inline test/common/quic/envoy_quic_client_session_test.cc:88]: should we simply remove this class, and have the mock factory do the custom creation?

File: test/common/quic/envoy_quic_server_session_test.cc [hunk 9]
```diff
@@ -122,6 +122,29 @@ class TestEnvoyQuicTlsServerHandshaker : public quic::TlsServerHandshaker,
   quic::QuicReferenceCountedPointer<quic::QuicCryptoNegotiatedParameters> params_;
 };
 
+class EnvoyQuicTestCryptoServerStreamFactory : public EnvoyQuicCryptoServerStreamFactoryInterface {
+public:
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override { return nullptr; }
+  std::string name() const override { return "quic.test_crypto_server_stream"; }
+
+  std::unique_ptr<quic::QuicCryptoServerStreamBase>
+  createEnvoyQuicCryptoServerStream(const quic::QuicCryptoServerConfig* crypto_config,
```
Comment (alyssawilk) [inline test/common/quic/envoy_quic_server_session_test.cc:131]: ditto here

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo a few nits.

@adisuissa can you do API review in Harvey's absence?

File: source/common/quic/client_connection_factory_impl.h (hunk 6 above)
Comment (danzh2010) [reply source/common/quic/client_connection_factory_impl.h:52]: This is in listener config now. And the extension config name has "Server" in it, and is also already documented to be used for server crypto stream.

Comment (alyssawilk) [reply source/common/quic/client_connection_factory_impl.h:52]: Ah sorry, maybe rephrase the comment then?
Hard code with the default crypto stream as there's no pluggable crypto for upstream Envoy?

File: test/common/quic/envoy_quic_client_session_test.cc (hunk 8 above)
Comment (danzh2010) [reply test/common/quic/envoy_quic_client_session_test.cc:88]: We don't have mock factories for crypto streams yet, and the mock factories probably will end up doing what these test classes do for now without more interfaces to mock.

File: api/envoy/extensions/quic/v3/crypto_stream.proto [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.quic.v3";
+option java_outer_classname = "CryptoStreamProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: QUIC server crypto stream config]
+// [#extension: quic.quiche_crypto_server_stream]
```
Comment (adisuissa) [inline api/envoy/extensions/quic/v3/crypto_stream.proto:13]: I think the extension name should start with `envoy.quic.server.crypto_stream` (the category defined above).

File: api/envoy/extensions/quic/v3/proof_source.proto [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.quic.v3";
+option java_outer_classname = "ProofSourceProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: QUIC proof source config]
+// [#extension: envoy.quic.filter_chain_proof_source]
```
Comment (adisuissa) [inline api/envoy/extensions/quic/v3/proof_source.proto:13]: I think the extension name should start with `envoy.quic.proof_source` (the category defined above).

File: api/envoy/config/listener/v3/quic_config.proto [hunk 12]
```diff
@@ -48,4 +50,14 @@ message QuicProtocolOptions {
   // bound by 6000, regardless of this field or how many connections there are.
   google.protobuf.UInt32Value packets_to_read_to_connection_count_ratio = 5
       [(validate.rules).uint32 = {gte: 1}];
+
+  // Configure which implementation of quic::QuicCryptoClientStreamBase to be used for this listener.
+  // If not specified the :ref:`QUICHE default one configured by <envoy_v3_api_msg_extensions.quic.v3.CryptoServerStreamConfig>` will be used.
+  // [#extension-category: envoy.quic.server.crypto_stream]
+  core.v3.TypedExtensionConfig crypto_stream = 6;
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:57]: Maybe append the suffix `_config` to this field and the one below.

**Review by adisuissa — COMMENTED** [review]
Left some API comments.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 12 above)
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto]: done

File: api/envoy/extensions/quic/v3/crypto_stream.proto (hunk 10 above)
Comment (danzh2010) [reply api/envoy/extensions/quic/v3/crypto_stream.proto]: done

**Review by alyssawilk — DISMISSED** [review]

**Comment by danzh2010** [comment]
ping @envoyproxy/api-shepherds

**Review by adisuissa — COMMENTED** [review]
API LGTM.
@htuch can you please take another look?

File: api/envoy/extensions/quic/v3/crypto_stream.proto [hunk 13]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
```
Comment (htuch) [inline api/envoy/extensions/quic/v3/crypto_stream.proto:1]: Actually, one tiny nit. I think both API and extensions should have distinct directories for each category of extension. So, there should be `api/envoy/extensions/quic/crypto_stream/v3` etc. (same in `source/extensions`)

**Review by htuch — DISMISSED** [review]

File: api/envoy/extensions/quic/v3/crypto_stream.proto (hunk 13 above)
Comment (danzh2010) [reply api/envoy/extensions/quic/v3/crypto_stream.proto]: done

**Review by adisuissa — COMMENTED** [review]
API looks good, but one question to verify that this is intentional:
Why do you set the extension name to be different than the extension path in both extensions (e.g., you use `envoy.quic.server.crypto_stream.quiche` instead of `envoy.quic.crypto_stream`)?

**Comment by danzh2010** [comment]
> API looks good, but one question to verify that this is intentional:
> Why do you set the extension name to be different than the extension path in both extensions (e.g., you use `envoy.quic.server.crypto_stream.quiche` instead of `envoy.quic.crypto_stream`)?

Yes, that's intentional. "server" and "quiche" is the unique to this extensions. We might add client or non-quiche-default extensions in the future.

**Comment by danzh2010** [comment]
> > API looks good, but one question to verify that this is intentional:
> > Why do you set the extension name to be different than the extension path in both extensions (e.g., you use `envoy.quic.server.crypto_stream.quiche` instead of `envoy.quic.crypto_stream`)?
> 
> Yes, that's intentional. "server" and "quiche" is the unique to this extensions. We might add client or non-quiche-default extensions in the future.

And I will fix their ordering as a follow up.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by danzh2010** [comment]
Can this be merged?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 14]
```diff
@@ -3,8 +3,10 @@ syntax = "proto3";
 package envoy.config.listener.v3;
 
 import "envoy/config/core/v3/base.proto";
+import "envoy/config/core/v3/extension.proto";
 import "envoy/config/core/v3/protocol.proto";
 
+import "google/protobuf/any.proto";
```
Comment (phlax) [inline api/envoy/config/listener/v3/quic_config.proto:9]: this is unused - could you remove it

File: api/envoy/config/listener/v3/quic_config.proto [hunk 15]
```diff
@@ -48,4 +50,14 @@ message QuicProtocolOptions {
   // bound by 6000, regardless of this field or how many connections there are.
   google.protobuf.UInt32Value packets_to_read_to_connection_count_ratio = 5
       [(validate.rules).uint32 = {gte: 1}];
+
+  // Configure which implementation of quic::QuicCryptoClientStreamBase to be used for this listener.
```
Comment (phlax) [inline api/envoy/config/listener/v3/quic_config.proto:54]: could you surround `quic::QuicCryptoClientStreamBase` with ` `` ` as its a letiral

File: api/envoy/config/listener/v3/quic_config.proto [hunk 16]
```diff
@@ -48,4 +50,14 @@ message QuicProtocolOptions {
...
+  // If not specified the :ref:`QUICHE default one configured by <envoy_v3_api_msg_extensions.quic.crypto_stream.v3.CryptoServerStreamConfig>` will be used.
+  // [#extension-category: envoy.quic.server.crypto_stream]
+  core.v3.TypedExtensionConfig crypto_stream_config = 6;
+
+  // Configure which implementation of quic::ProofSource to be used for this listener.
```
Comment (phlax) [inline api/envoy/config/listener/v3/quic_config.proto:59]: please make `quic::ProofSource` a literal also

**Review by phlax — CHANGES_REQUESTED** [review]
@danzh2010 ive added a couple of docs nits

**Review by htuch — COMMENTED** [review]
/lgtm api

**Review by alyssawilk — APPROVED** [review]
LGTM!  Looks like CI flaked though - I'll kick off another run but feel free to ping me if you notice it's green before I do :-)
